File: 2020/2020W22/PreppindataW22.py
import pandas as pd
import random

xlsx = pd.ExcelFile("F:/Data/2020W22 Input.xlsx")

ttl = pd.read_excel(xlsx,'Total Market')
comp = pd.read_excel(xlsx,'Companies')
scent = pd.read_excel(xlsx,'Scents')

march_ttl = ttl['March Sales'][0]
april_ttl = march_ttl*(1+ttl['Growth'][0])

comp['random1'] = comp.apply(lambda x:random.random(),axis=1)
comp['random2'] = comp.apply(lambda x:random.random(),axis=1)
comp['March'] = march_ttl*comp['random1']/comp['random1'].sum()
comp['randomRank'] = comp['random2'].rank().astype(int)
comp['ChangeBps'] = comp['randomRank'].apply(lambda x: -30+x*10)
comp['April'] = april_ttl*(comp['random1']/comp['random1'].sum()+comp['ChangeBps']/10000)

final = comp[['Company','March','April']].melt(id_vars='Company', value_name='CM Sales', var_name='Month')
final['dummy'] = 0
scent['dummy'] = 0
final = pd.merge(final, scent, on='dummy')

final['random1'] = final.apply(lambda x:random.random(),axis=1)
final['randomttl'] = final['random1'].groupby([final['Company'],final['Month']]).transform('sum')
final['Sales'] = final.apply(lambda x: x['CM Sales']*x['random1']/x['randomttl'],axis=1)
final = final[['Company', 'Month', 'Soap Scent', 'Sales']]

File: 2020/2020W22/preppindataw22.R
library(dplyr)
library(readxl)
library(tidyr)

xlsx <- "F:/Data/2020W22 Input.xlsx"

ttl <- read_excel(xlsx, sheet = "Total Market") 
comp <- read_excel(xlsx, sheet = "Companies") 
scent <- read_excel(xlsx, sheet = "Scents") 

march_ttl <- ttl$'March Sales'[1]
april_ttl <- march_ttl*(1+ttl$'Growth'[1])

final <- comp %>%
  rowwise() %>%
  mutate('random1' = runif(1), 
         'random2' = runif(1)) %>%
  ungroup() %>%
  mutate('March' = march_ttl*`random1`/ sum(`random1`),
         'ChangeBps' = -30+rank(desc(as.numeric(`random2`)))*10,
         'April' = april_ttl*(`random1`/ sum(`random1`)+`ChangeBps`/10000)) %>%
  select(c('Company','March','April')) %>%
  pivot_longer(., cols=matches('^(March|April)'), names_to='Month', values_to='CM Sales') %>%
  merge(scent, all=TRUE) %>%
  rowwise() %>%
  mutate('random1' = runif(1)) %>%
  ungroup() %>%
  group_by(`Company`,`Month`) %>%
  mutate('Sales' = `CM Sales`*`random1`/sum(`random1`)) %>%
  arrange(`Company`,`Month`) %>%
  select(c('Company', 'Month', 'Soap Scent', 'Sales'))

View(final)